### PR TITLE
File Store & Link Generation

### LINKED ISSUE #96: Saved Map JSON Upload and Store
After issue #95 is complete, this JSON data should be uploaded to the pocketbase instance.

### PR DESCRIPTION
**Pull request by Fish1**
# Feature
The share modal can now generate a shareable link.

[Video.webm](https://user-images.githubusercontent.com/13600696/217134127-8ffe76c3-9e2b-4e74-8005-868229b9a431.webm)

# Code

Collections - Added a maps collection that stores the map json, and a hash of the map.

Pocketbase API - Added an API route that will gzip a json file and store it. It will only store the map if the hash of the file doesn't already exist.

Client - The client will display a popup if the user tries to open a poorly formatted map.

Client - The share map modal has a generate link button. When the link is generated from the server it will provide an easy to use button that will copy the link to the users clipboard.

closes #96

### CODE REVIEW

File: apps/yapms/src/lib/components/modals/sharemodal/ShareModal.svelte [hunk 1]
```diff
@@ -3,18 +3,64 @@
...
+  async function generateLink() {
+    const blob = new Blob(
+      [JSON.stringify(generateJson())],
+      { type: "application/json" }
+    );
+
+    const formData = new FormData();
+    formData.append("data", blob, "data.json");
+
+    const url = new URL("http://localhost:8080/api/yapms.com/map");
```
Comment (Flizzyclone) [inline apps/yapms/src/lib/components/modals/sharemodal/ShareModal.svelte]: Probably want to change localhost:8080 to an environment variable

File: apps/yapms/src/routes/app/+layout.svelte [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	onMount(async () => {
+		const map = encodeURIComponent($page.url.searchParams.get("m") ?? "");
+		// if the user is not loading a map, redirect them to the default map
+		if (map === "") {
+			await goto(`/app/usa/presidential/2022`);
+			return;
+		}
+
+		// load the requested map from pocketbase
+		const data = await fetch(`http://localhost:8080/api/files/maps/${map}/data.json.gz`);
```
Comment (Flizzyclone) [inline apps/yapms/src/routes/app/+layout.svelte]: Same thing here with the environment variable

File: apps/yapms/src/routes/app/+page.svelte [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<script lang="ts">
+  import { page } from "$app/stores";
+  const mapID = $page.url.searchParams.get("map");
```
Comment (Flizzyclone) [inline apps/yapms/src/routes/app/+page.svelte]: My understanding here is that under most circumstances this searchParam wouldn't exist and instead the param would be "m". Perhaps something that got changed halfway through?

**Review by Flizzyclone — APPROVED** [review]
This all looks good! Just left some notes on some things I wasn't quite sure about/think may be an improvement when actually deploying the app.

File: apps/yapms/src/lib/components/modals/sharemodal/ShareModal.svelte (hunk 1 above)
Comment (Fish1) [reply apps/yapms/src/lib/components/modals/sharemodal/ShareModal.svelte]: Fixed

File: apps/yapms/src/routes/app/+layout.svelte (hunk 2 above)
Comment (Fish1) [reply apps/yapms/src/routes/app/+layout.svelte]: Fixed

File: apps/yapms/src/routes/app/+page.svelte (hunk 3 above)
Comment (Fish1) [reply apps/yapms/src/routes/app/+page.svelte]: Yeah, this was just me messing around with something else. I've deleted it now, for this PR.

**Comment by Fish1** [comment]
Thanks for those callouts. They should all be environment variables now.

**Comment by Flizzyclone** [comment]
LGTM, thanks for implementing these changes!

**Review by Flizzyclone — APPROVED** [review]

**Pull request merged**